src/powerset.rs

```diff
@@ -0,0 +1,149 @@
+use std::{fmt, mem};
+use std::usize;
+use alloc::vec::Vec;
+
+use super::combinations::Combinations;
+use super::lazy_buffer::LazyBuffer;
+use super::size_hint;
+
+/// An iterator over the powerset of all elements from its source iterator.
+///
+/// See [`.powerset()`](../trait.Itertools.html#method.powerset) for more
+/// information.
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct Powerset<I: Iterator> {
+    inner: Inner<I>,
+    // Position from start of iteration, used for first element and size-hint.
+    pos: usize,
+}
+
+impl<I> Clone for Powerset<I>
+    where I: Clone + Iterator,
+          I::Item: Clone,
+{
+    clone_fields!(inner, pos);
+}
+
+impl<I> fmt::Debug for Powerset<I>
+    where I: Iterator + fmt::Debug,
+          I::Item: fmt::Debug,
+{
+    debug_fmt_fields!(Powerset, inner, pos);
+}
+
+/// Create a new `Powerset` from a cloneable iterator.
+pub fn powerset<I>(src: I) -> Powerset<I>
+    where I: Iterator,
+          I::Item: Clone,
+{
+    Powerset {
+        inner: Inner::Buffer(src, alloc::vec![]),
+        pos: 0,
+    }
+}
+
+// Represents the inner state of a `Powerset` iterator.
+#[derive(Clone, Debug)]
+enum Inner<I: Iterator> {
+    // Buffering mode, source iterator still yielding elements.
+    Buffer(I, Vec<I::Item>),
+    // Combination mode, yielding combinations of buffered elements.
+    Combs(Combinations<I>),
+    // Temporary state for in-place switching of Buffer->Combs.
+    Empty_,
+}
```

Comment:
Couldn't we just always have a `Combinations`, starting with `k=0`, and increment `k` after the `k`-combinations are exhausted, until we reach `k==n+1`?

This would obviate many of the case distinctions on `inner`, I guess.

Or is there a reason why this `enum`/case distinction is strictly necessary?

Reply:
Yes, I was torn on this and went with the 'inner' approach due to a performance benefit for use cases involving lots of small powersets. I think you're right, it's probably not worth the extra complexity.